src/Lecture3.hs

```diff
@@ -229,24 +253,29 @@ types that can have such an instance.
 -- instance Functor Weekday where
 -- instance Functor Gold where
 -- instance Functor Reward where
--- instance Functor List1 where
--- instance Functor Treasure where
-
-{- | Functions are first-class values in Haskell. This means that they
-can be even stored inside other data types as well!
-
-Now, you have a function inside some 'Functor'. You're a given an
-element and you need to apply the function inside the 'Functor' to a
-given element.
-
->>> apply 5 (Just (+ 3))
-Just 8
->>> apply 5 Nothing
-Nothing
->>> apply [1 .. 10] (Just (drop 7))
-Just [8,9,10]
->>> apply 5 [(+ 3), (* 4), div 17]
-[8,20,3]
-
--}
-apply = error "TODO"
+instance Functor List1 where
+  fmap :: (a -> b) -> List1 a -> List1 b
+  fmap f (List1 val l) = List1 (f val) (map f l)
+
+instance Functor Treasure where
+  fmap :: (a -> b) -> Treasure a -> Treasure b
+  fmap _ NoTreasure = NoTreasure
+  fmap f (SomeTreasure x) = SomeTreasure (f x)
+
+-- | Functions are first-class values in Haskell. This means that they
+-- can be even stored inside other data types as well!
+--
+-- Now, you have a function inside some 'Functor'. You're a given an
+-- element and you need to apply the function inside the 'Functor' to a
+-- given element.
+--
+-- >>> apply 5 (Just (+ 3))
+-- Just 8
+-- >>> apply 5 Nothing
+-- Nothing
+-- >>> apply [1 .. 10] (Just (drop 7))
+-- Just [8,9,10]
+-- >>> apply 5 [(+ 3), (* 4), div 17]
+-- [8,20,3]
+apply :: Functor f => a -> f (a -> a) -> f a
```

Comment:
That's correct! Although, you can make the type even more polymorphic because the function doesn't need to return the same type that it gets.

Probably it wasn't clear from the examples, so this is something that can be improved in the course itself 🙂 

```suggestion
apply :: Functor f => a -> f (a -> b) -> f b
```